File: tictactoe.go
package main

import(
	"fmt"
	"bufio"
	"os"
	"strconv"
	"log"
	"strings"
)


type Case struct{
	content string
}

func (c *Case) changecontent(ncont string) {
	c.content = ncont
}
var caselist [9]Case

type Player struct {
	name string
	sign string
}

func (p Player) play() 	{
	scanner := bufio.NewReader(os.Stdin)
	printboard(caselist)
	fmt.Println("")
	fmt.Println(p.name, "Your turn !")
	fmt.Println("Choose a number to play between 123456789")
	text, _ := scanner.ReadString('\n')
	trans, err := strconv.Atoi(strings.Replace(text, "\r\n", "",  1))
	if err != nil {
		log.Fatal(err)
	}
	trans = trans - 1
	switch caselist[trans].content {
		case "#" :
			caselist[trans].changecontent(p.sign)
		default : 
			fmt.Println("Play something else")
			p.play()
	}
}



func printboard(cselist [9]Case) {
	fmt.Printf(cselist[0].content + "|" + cselist[1].content + "|" + cselist[2].content + "\n" + "-----" + "\n" + cselist[3].content + "|" + cselist[4].content + "|" + cselist[5].content + "\n" + "-----" + "\n" + cselist[6].content + "|" + cselist[7].content + "|" + cselist[8].content)
}

func initcases() {
	for i:=0; i!=9; i++  {
		caselist[i] = Case{"#"}
	}
}

func getstate() string{
	if caselist[0].content == caselist[1].content && caselist[1].content == caselist[2].content && caselist[0].content != "#" {
		return(caselist[0].content)
	}
	if caselist[3].content == caselist[4].content && caselist[4].content == caselist[5].content && caselist[3].content != "#" {
		return(caselist[3].content)
	}
	if caselist[6].content == caselist[7].content && caselist[7].content == caselist[8].content && caselist[6].content != "#" {
		return(caselist[6].content)
	}
	if caselist[0].content == caselist[3].content && caselist[3].content == caselist[6].content && caselist[0].content != "#" {
		return(caselist[0].content)
	}
	if caselist[1].content == caselist[4].content && caselist[4].content == caselist[7].content && caselist[1].content != "#" {
		return(caselist[1].content)
	}
	if caselist[2].content == caselist[5].content && caselist[5].content == caselist[8].content && caselist[2].content != "#" {
		return(caselist[2].content)
	}
	if caselist[0].content == caselist[4].content && caselist[4].content == caselist[8].content && caselist[0].content != "#" {
		return(caselist[0].content)
	}
	if caselist[2].content == caselist[4].content && caselist[4].content == caselist[6].content && caselist[2].content != "#" {
		return(caselist[2].content)
	}
	for i:=0; i !=9; i++ {
		if caselist[i].content == "#" {
			return("NOTEND")
		}
	}
	return("DRAW")
}


func main() {
	initcases()
	Jhon := Player{"Jhon", "X"}
	Jack := Player{"Jack", "O"}
	for true {
		if getstate() == "NOTEND" {
			Jhon.play()
			for i:=0; i<100; i++ {
				fmt.Println("\n")
			}
		}
		if getstate() == "NOTEND" {
			Jack.play()			
			for i:=0; i<100; i++ {
				fmt.Println("\n")
			}
		}
		if getstate() != "NOTEND" {
			break
		}
	}
	switch getstate() {
	case "DRAW" :
		fmt.Println("This is a draw !")
	case "O" :
		fmt.Println("Jack won the game !")
	case "X" :
		fmt.Println("Jhon won the game !")
	}
}

File: tictactoe.py
import sys

class Player() :
	def __init__(self,sign,name) :
		self.sign = sign
		self.name = name
	def play(self,case) :
		if case.occ != "#" :
			print("Case is already taken, consider playing something else")
			return(False)
		else :
			case.fill(self.sign)
			return(True)
class Case() :
	def __init__(self):
	 	self.occ = "#"
	def fill(self,sign) :
	 	self.occ = sign

def initcases() :
	caselist = []
	for i in range(9) :
		caselist.append(Case())
	return(caselist)

def iswon(p,cl) :
	for i in range(0,7,3) :
		if cl[0+i].occ == cl[1+i].occ ==cl[2+i].occ == p.sign :
			return(True)
	for i in range(0,3,1) :
		if cl[0+i].occ == cl[3+i].occ == cl[6+i].occ == p.sign :
			return(True)
	if cl[0].occ == cl[4].occ == cl[8].occ == p.sign :
		return(True)
	if cl[2].occ == cl[4].occ == cl[6].occ == p.sign  :
		return(True)
	return(False)

def isdraw(cl) :
	for case in cl :
		if case.occ == "#" :
			return(False)
	return(True)


def display(cl) :
	print(cl[0].occ+"|"+cl[1].occ+"|"+cl[2].occ+"\n-----\n"+cl[3].occ+"|"+cl[4].occ+"|"+cl[5].occ+"\n-----\n"+cl[6].occ+"|"+cl[7].occ+"|"+cl[8].occ+"\n")

def game() :
	Jhon = Player("X","Jhon")
	Jack = Player("O","Jack")
	cl = initcases()
	while True :
		if iswon(Jhon,cl) :
			print("Jhon won the game !")
			sys.exit()
		if iswon(Jack,cl) :
			print("Jack won the game !")
			sys.exit()
		if isdraw(cl) :
			print("This is a draw :(")
			sys.exit()
		display(cl)
		while not Jhon.play(cl[int(input("Your turn, Jhon ! play 123456789\n"))-1]) :
			pass
		print("\n\n\n\n")
		display(cl)
		if iswon(Jhon,cl) :
			print("Jhon won the game !")
			sys.exit()
		if iswon(Jack,cl) :
			print("Jack won the game !")
			sys.exit()
		if isdraw(cl) :
			print("This is a draw :(")
			sys.exit()
		while not Jack.play(cl[int(input("Your turn, Jack ! play 123456789\n"))-1]) :
			pass
		print("\n\n\n\n")

if __name__ == '__main__' :
	game()
